File: 25.K 个一组翻转链表_辅助数组.cc
#include <vector>
using std::vector;

// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        ListNode* dummy = new ListNode(0);
        dummy->next = head;
        ListNode* last = dummy;
        vector<ListNode*> vec(k);
        int cur = 0;
        while (head != nullptr) {
            vec[cur] = head;
            ++cur;
            head = head->next;
            if (cur == k) {
                vec[0]->next = nullptr;
                for (int i = 1; i < k; ++i) {
                    vec[i]->next = vec[i-1];
                }
                last->next = vec[k-1];
                last = vec[0];
                vec.clear();
                cur = 0;
            }
        }
        if (cur != 0) {
            last->next = vec[0];
            vec.clear();
        }
        return dummy->next;
    }
};

File: Python/25.K 个一组翻转链表 辅助数组.py
from typing import Optional, List

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        dummy: ListNode = ListNode(0, head)
        last: ListNode = dummy
        arr: List[ListNode] = [None] * k
        cur: int = 0
        while head is not None:
            arr[cur] = head
            cur += 1
            head = head.next
            if cur == k:
                arr[0].next = None
                for i in range(1, k):
                    arr[i].next = arr[i-1]
                last.next = arr[k-1]
                last = arr[0]
                arr = [None] * k
                cur = 0
        if cur != 0:
            last.next = arr[0]
            del arr
        return dummy.next
